opentelemetry-sdk/src/trace/sampler.rs

```diff
@@ -13,6 +13,8 @@ pub use jaeger_remote::{JaegerRemoteSampler, JaegerRemoteSamplerBuilder};
 #[cfg(feature = "jaeger_remote_sampler")]
 use opentelemetry_http::HttpClient;
 
+const TRACE_FLAG_DEFERRED: TraceFlags = TraceFlags::new(0x02);
```

Comment:
https://w3c.github.io/trace-context/#trace-flags
I don't see such a flag definition in the official spec...

Reply:
It would've been better to provide a way to apply sampling inside a propagator when extracting tracing context.
But that's a separate thing that would require a major release.

Reply:
I think it would be best to merge my fix now and then see how this `TRACE_FLAG_DEFERRED` solution could be refactored for the next major release

Reply:
I think it's unfortunate that the `B3` code is using `TraceFlags` to send over this information between `Extractor` and `Injector`. If your application is set up to accept `B3` headers and inject both `B3` and `w3c trace context` headers, you will end up with _illegal_ trace flags in the `w3c` headers. Also, the sampling decision made here will not propagate correctly to the outgoing `B3` headers since they look at the `deferred` flag first, and skip the sampling decision.

To support this use case, I think it would be better to define a separate `struct` to hold this information (and not `B3` specific) in the API or SDK that is added as to the `Context` and then read by the `Sampler`. The `B3` code would need to be updated as well.

Reply:
I testes this patch and it works correctly including context propagation.

Propagation works correctly because existing propagators clear all flags except for `TraceFlags::SAMPLED` right now, i.e.:
https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/propagation/trace_context.rs#L136

I agree that going forward it would be best to store this information separately instead of using trace flags, though.

Reply:
> Also, the sampling decision made here will not propagate correctly to the outgoing B3 headers since they look at the deferred flag first, and skip the sampling decision.

Hm... you may be right. I should check that. So far I only verified that `w3c` headers are generated correctly.